README.md

```diff
@@ -156,40 +158,47 @@ Finalizers can help prevent memory leaks. [Comlink](https://github.com/GoogleChr
 
 ## Using `WeakRef`s and `FinalizationGroup`s together
 
-`WeakRef` and `FinalizationGroup` often make sense to use together. There are several kinds of data structures that want to weakly point to a value, and do some kind of cleanup when that value goes away.
+It sometimes makes sense to use `WeakRef` and `FinalizationGroup` together. There are several kinds of data structures that want to weakly point to a value, and do some kind of cleanup when that value goes away.
 
-### Clearing the unused cache keys
+### Weak caches
 
-In the initial example from this README, a cache can use `WeakRef` to point to large `ArrayBuffer`s to allow them to be garbage collected, but still reused when available. To allow the cache *keys* to be deleted when they are no longer in use, a `FinalizationGroup` can register a callback to delete the `Map` entries, as follows:
+In the initial example from this README, `makeWeakCached` used a `Map` whose values were wrapped in `WeakRef` instances.  This allowed the cached values to be collected, but leaked memory in the form of the entries in the map.  A more complete version of `makeWeakCached` uses finalizers to fix this memory leak.
 
 ```js
-const cache = new Map();
-const finalizationGroup = new FinalizationGroup(iterator => {
-  for (const name of iterator) {
-    const ref = cache.get(name);
-    if (ref !== undefined) {
-      if (ref.deref() === undefined) {
-        cache.delete(name);
-      }
+// Fixed version that doesn't leak memory.
+function makeWeakCached(f) {
+  const cache = new Map();
+  const cleanup = new FinalizationGroup(iterator => {
+    for (const key of iterator) {
+      // See note below on concurrency considerations.
+      const ref = cache.get(key);
+      if (ref && !ref.deref()) cache.delete(key);
     }
-  }
-});
-function getImageCached(name) {
-  let ref = cache.get(name);
-  if (ref !== undefined) {
-    let deref = ref.deref();
-    if (deref !== undefined) {
-      return deref;
+  });
+
+  return key => {
+    const ref = cache.get(key);
+    if (ref) {
+      const cached = ref.deref();
+      // See note below on concurrency considerations.
+      if (cached !== undefined) return cached;
```

Comment:
Another approach would be to call `cleanup.unregister(key)` if `cached === undefined` here. That way we're guaranteed the `key` won't be in the cleanup callback anymore, until our new cached object is collected that is.

Reply:
This would keep the consistency work in a single location, and greatly simplify the 1. bullet in the explainer below.